torch/autograd/gradcheck.py

```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
         jacobian_cols[d_idx] = _compute_numerical_jvps_wrt_specific_input(jvp_fn, eps, x.is_complex())
     return _combine_jacobian_cols(jacobian_cols, outputs, input, input.numel())
 
+def _get_analytical_jacobian_forward_ad(fn, inputs, outputs, *, all_u=None) -> Tuple[Tuple[torch.Tensor, ...], ...]:
+    """Computes the analytical Jacobian using forward mode AD of `fn(inputs)` with respect
+    to `target`. Returns M * N Jacobians where N is the number of tensors in target that require grad and
+    M is the number of non-integral outputs.
+    Contrary to other functions here, this function requires "inputs" to actually be used by the function.
+    The computed value is expected to be wrong if the function captures the inputs by side effect instead of
+    using the passed ones (many torch.nn tests do this).
+
+    Args:
+        fn: the function to compute the jacobian for
+        inputs: inputs to `fn`
+        outputs: provide precomputed outputs to avoid one extra invocation of fn
+        all_u (optional): if provided, the Jacobian will be right multiplied with this vector
+
+    Returns:
+        A list of M N-tuples of tensors
+    """
+    # To avoid early import issues
+    fwAD = torch.autograd.forward_ad
+
+    if any(i.is_complex() for i in inputs):
+        raise ValueError("Expected inputs to be non-complex for _get_analytical_jacobian_forward_ad.")
```

Comment:
We can probably add some tests here to check if all of these errors can be reached
ditto for the one below for MKLDNN inputs

Reply:
The output validation is shared, so no need to test anything. But adding a test for this input check indeed.